ingestion-edge/ingestion_edge/publish.py

```diff
@@ -3,44 +3,135 @@
 # file, you can obtain one at http://mozilla.org/MPL/2.0/.
 
 from .conf import ROUTE_TABLE
+from .conf import QUEUE_PATH
 from datetime import datetime
 from flask import Blueprint, request
-from google.cloud.pubsub import PublisherClient
+from google.api_core.exceptions import Aborted
+from google.api_core.exceptions import Cancelled
+from google.api_core.exceptions import Forbidden
+from google.api_core.exceptions import RetryError
+from google.api_core.exceptions import ServerError
+from google.api_core.exceptions import TooManyRequests
+from google.cloud.pubsub_v1 import PublisherClient
+from google.cloud.pubsub_v1.publisher.exceptions import PublishError
+from google.cloud.pubsub_v1.types import PubsubMessage
+from persistqueue import SQLiteAckQueue
+from persistqueue.exceptions import Empty
 from typing import Tuple
+import json
+
+FLUSH_EMPTY_STATUS = 204
+FLUSH_COMPLETE_STATUS = 200
+FLUSH_INCOMPLETE_STATUS = 504
+# Errors where message should be queued locally after pubsub delivery fails
+# To minimize data loss include errors that may require manual intervention
+TRANSIENT_ERRORS = (
+    # The API thinks the client should retry
+    Aborted,
+    # The API was interrupted when handling the request
+    Cancelled,
+    # The PublisherClient lacks permissions and may require manual intervention
+    Forbidden,
+    # API call ran out of retries
+    RetryError,
+    # API outage or connection issue
+    ServerError,
+    # A python Future timed out
+    TimeoutError,
+    # The API is throttling requests
+    TooManyRequests,
+)
 
 publish = Blueprint('publish', __name__)
 client = PublisherClient()
+q = SQLiteAckQueue(QUEUE_PATH)
+
+
+def _publish(topic, data, attrs) -> str:
+    """Publish one message to the topic
+
+    Uses PublisherClient to try to publish the message in a batch and handles
+    non-transient batch failure by retrying synchronously with a single message
+    """
+    try:
+        return client.publish(topic, data, **attrs).result()
```

Comment:
I want to make sure I understand this correctly. We have a `client` defined in the module scope. One instance of this Flask app will have many threads running and all those threads will be accessing the same client. When a thread hits this line, it blocks waiting for the client to complete and release a batch. So a batch will contain at most n messages where n is the number of Flask application threads. A single thread is only ever handling one message at a time; it never moves on to the next message until the pubsub client acknowledges delivery.

It may be worth writing up documentation on the semantics here, either in a docstring here or in a spot where we configure batching and threading settings.

Reply:
TL;DR no it's asynchronous.

Note: my understanding here is based on very recent reading, and may be more detailed than you need, but it's all new information for me.

> We have a client defined in the module scope. 

Yes

> One instance of this Flask app will have many threads running and all those threads will be accessing the same client.

Not necessarily. With gunicorn each asynchronous worker (may be a process or thread) will have one instance of this Flask app. Each worker will have many coroutines. All of those coroutines will be accessing the same client. Coroutines differ from threads in that they are managed by python (or whatever language is using them) so they only get paused when python thinks it's okay, like when waiting on a `concurrent.futures.Future` or IO operation.

Note: our docker container is currently using the `sync` worker class by default. This will be changed to an asynchronous worker class like `gevent` in the future.

> When a thread hits this line, it blocks waiting for the client to complete and release a batch.

When a coroutine hits this line it should yields because `publish(...)` returns a `Future` which the coroutine waits for via `.result()`.

> So a batch will contain at most n messages where n is the number of Flask application threads.

Where n is the number of coroutines being run by this worker, and a worker will run a coroutine for every request it receives. The actual limit is lower because the client library also starts new batches based on limits for timeout, message count, and batch size.

> A single thread is only ever handling one message at a time; it never moves on to the next message until the pubsub client acknowledges delivery.

A single coroutine is only ever handling one message at a time, but a single worker may handle thousands of connections, each of which may contain multiple requests.

Reply:
> configure batching and threading settings

we should modify the ~~batching~~ client settings eventually because it defaults to a 60 second timeout on api calls, and we may call the api twice per request before queuing to disk, and I suspect that kind of latency is too high.

~~threading~~ worker settings are managed via env vars by ops. We should eventually add a section to each `README.md` indicating what settings exist and how they are set in stage/prod.

I think we should handle both of those in later PRs